docs/internal/DistributedArchitectureGuide.md

```diff
@@ -365,18 +365,142 @@ There are several more Decider Services, implementing the `AutoscalingDeciderSer
+
+Some long-running tasks are implemented to be cancel-able. Cancellation can be done via the [Cancel Task REST API] or programmatically using [TaskManager#cancelTaskAndDescendants].
+
+When a task is cancelled, that task and all of its descendant tasks will be cancelled. In order to support cancellation, the [Task] instance associated with the task must extend [CancellableTask].
+
+When a [Task] extends [CancellableTask] the [TaskManager] keeps track of it and any child tasks that it spawns. When the task is cancelled, requests are sent to any nodes that have had child tasks submitted to them to ban the starting of any further children of that task, and any cancellable child tasks already running are themselves cancelled (see [BanParentRequestHandler]).
+
+When a cancellable task dispatches child requests through the [TransportService], it registers a proxy so that if one of those child requests completes exceptionally, all other outstanding child requests descended from the same parent are cancelled (see [UnregisterChildTransportResponseHandler]).
+
+It is the job of any long-running workload tracked by a [CancellableTask] to periodically check whether it has been cancelled and if so, halt.
+
+### Publishing Task Results
+
+[TaskResult]:https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/tasks/TaskResult.java
+[TaskResultsService]:https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/tasks/TaskResultsService.java
+[CAT]:https://www.elastic.co/guide/en/elasticsearch/reference/current/cat.html
+[ActionRequest]:https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/ActionRequest.java
+[ActionRequest#getShouldStoreResult]:https://github.com/elastic/elasticsearch/blob/b633fe1ccb67f7dbf460cdc087eb60ae212a472a/server/src/main/java/org/elasticsearch/action/ActionRequest.java#L32
+[TaskResultStoringActionListener]:https://github.com/elastic/elasticsearch/blob/b633fe1ccb67f7dbf460cdc087eb60ae212a472a/server/src/main/java/org/elasticsearch/action/support/TransportAction.java#L149
+
+A list of tasks currently running in a cluster can be requested via the [Task management API], or the [cat task management API]. The former returns each task represented using the [TaskResult] DTO, the latter returning a more compact [CAT] representation.
+
+Some [ActionRequest]s allow the results of the actions they spawn to be stored upon completion. If [ActionRequest#getShouldStoreResult] returns true, a [TaskResultStoringActionListener] will be inserted into the chain of response listeners. [TaskResultStoringActionListener] serializes the [TaskResult] of the [TransportAction] and persists it in the `.tasks` index using the [TaskResultsService].
+
+The [Task management API] also exposes an endpoint where a task ID can be specified, this form of the API will return currently running tasks, or completed tasks whose results were persisted.
+
```

Comment:
I have a few questions around this idea of storing the task result.
1. Is the task result persisted in the .tasks index forever or do we cleanup the .tasks index periodically by removing the TaskResults that belong to tasks that were executed long ago ?
2. Do we persist the results of all tasks submitted by clients in this .task index ? If not, how can we accurately convey to the user the result of a task when the user queries using the "GET _tasks/<task_id>" endpoint ?
3. Where does the .task index live ? I guess my question is where does any system index live ? Does it exist only on the master eligible nodes ?
4. You do mention persistent tasks later on. But just to make sure I understand this correctly, the tasks we are talking about here, which are not Persistent tasks will be lost if a node is shutdown or fails while the task is ongoing ?

Reply:
Hi Ankita, I'll answer what I can in-line

> 1. Is the task result persisted in the .tasks index forever or do we cleanup the .tasks index periodically by removing the TaskResults that belong to tasks that were executed long ago ?

That is a good question and I don't know the answer. I'll skim through and see if there are any processes around that worth documenting.

> 2. Do we persist the results of all tasks submitted by clients in this .task index ? If not, how can we accurately convey to the user the result of a task when the user queries using the "GET _tasks/<task_id>" endpoint ?

I should be clearer about that. The `error` and `response` fields in the `TaskResult` will only ever be populated for completed tasks that were loaded from the index, otherwise they will be null. The task's status is included in the `TaskInfo` however and I imagine that's more commonly used to convey information about the state of a running task. I will improve this section.

> 3. Where does the .task index live ? I guess my question is where does any system index live ? Does it exist only on the master eligible nodes ?

I believe it's effectively a normal index. As I understand it, system indices are normal in every respect (sharded across the cluster, and updated in the normal ways), but there are now some limitations on which users can see and interact with them.

> 4. You do mention persistent tasks later on. But just to make sure I understand this correctly, the tasks we are talking about here, which are not Persistent tasks will be lost if a node is shutdown or fails while the task is ongoing ?

Correct, persistent tasks are a separate concept to stored results. Tasks that store their results are not necessarily persistent, but persistent tasks could be configured to store their results.

Reply:
> Do we persist the results of all tasks submitted by clients in this .task index ? If not, how can we accurately convey to the user the result of a task when the user queries using the "GET _tasks/<task_id>" endpoint ?

I attempted to make this clearer in 87db9c65373d9b24bd0ee8568bab156f1ac896ef

> Is the task result persisted in the .tasks index forever or do we cleanup the .tasks index periodically by removing the TaskResults that belong to tasks that were executed long ago ?

I've looked around and there doesn't seem to be any code for doing cleanup of the `.tasks` index, or documentation around it. I assume this must need to be done manually? I imagine it's fairly low volume though, it doesn't look like many actions allow storing of results and when they do, they only seem to do it when `wait_for_completion` is false.